sdk/attestation/azure-security-attestation/src/attestation_client.cpp

```diff
@@ -0,0 +1,158 @@
+      std::move(perRetrypolicies),
+      std::move(perCallpolicies));
+}
+
+Azure::Response<AttestationOpenIdMetadata> AttestationClient::GetOpenIdMetadata(
+    Azure::Core::Context const& context) const
+{
+  auto request = AttestationCommonRequest::CreateRequest(
+      m_endpoint, HttpMethod::Get, {".well-known/openid-configuration"}, nullptr);
+
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+  auto openIdMetadata(OpenIdMetadataSerializer::Deserialize(response));
+  return Response<AttestationOpenIdMetadata>(std::move(openIdMetadata), std::move(response));
+}
+
+Azure::Response<std::vector<AttestationSigner>>
+AttestationClient::GetAttestationSigningCertificates(Azure::Core::Context const& context) const
+{
+  auto request
+      = AttestationCommonRequest::CreateRequest(m_endpoint, HttpMethod::Get, {"certs"}, nullptr);
+
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+  auto jsonWebKeySet(JsonWebKeySetSerializer::Deserialize(response));
+  std::vector<Models::AttestationSigner> signers;
+  for (const auto& jwk : jsonWebKeySet.Keys)
+  {
+    AttestationSignerInternal internalSigner(jwk);
+    signers.push_back(internalSigner);
+  }
+  return Response<std::vector<AttestationSigner>>(signers, std::move(response));
+}
+
+Azure::Response<AttestationToken<AttestationResult>> AttestationClient::AttestSgxEnclave(
+    std::vector<uint8_t> const& sgxQuote,
+    AttestOptions options,
+    Azure::Core::Context const& context) const
+{
+  AttestSgxEnclaveRequest attestRequest{
+      sgxQuote,
+      options.InittimeData,
+      options.RuntimeData,
+      options.DraftPolicyForAttestation,
+      options.Nonce};
+
+  std::string serializedRequest(AttestSgxEnclaveRequestSerializer::Serialize(attestRequest));
+
+  auto encodedVector = std::vector<uint8_t>(serializedRequest.begin(), serializedRequest.end());
+  Azure::Core::IO::MemoryBodyStream stream(encodedVector);
+  auto request = AttestationCommonRequest::CreateRequest(
+      m_endpoint, m_apiVersion, HttpMethod::Post, {"attest/SgxEnclave"}, &stream);
+
+  // Send the request to the service.
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+
+  // Deserialize the Service response token and return the JSON web token returned by the service.
+  std::string responseToken = AttestationServiceTokenResponseSerializer::Deserialize(response);
+
+  // Parse the JWT returned by the attestation service.
+  auto token
+      = AttestationTokenInternal<AttestationResult, AttestationResultSerializer>(responseToken);
+
+  // Validate the token returned by the service. Use the cached attestation signers in the
+  // validation.
+  CacheAttestationSigners(context);
+  token.ValidateToken(
+      options.TokenValidationOptions.HasValue() ? options.TokenValidationOptions.Value()
+                                                : this->m_tokenValidationOptions,
+      m_attestationSigners);
```

Comment:
Why do we need to validate here? Is that something that the Service ask customers to do as per documentation?  or is this just the SDK trying to exceed expectations?

Reply:
We should probably discuss this offline. Essentially the answer is: The client needs to be able to ensure that the token returned from the attestation service was actually generated by an SGX enclave contained within the attestation service. The token validation options defines a set of checks which are used to validate the token. There's one step that cannot be performed by the SDK (for <reasons>), that's the purpose of the client callback which allows customers to perform additional validations beyond the basic validations performed by the service.